mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -13,25 +17,104 @@ impl S3CrtClient {
         let mut message = self
-            .new_request_template("PUT", bucket)
+            .client
+            .new_request_template("PUT", &self.bucket)
             .map_err(S3RequestError::construction_failure)?;
 
-        let key = format!("/{key}");
+        let key = format!("/{}", self.key);
         message
             .set_request_path(&key)
             .map_err(S3RequestError::construction_failure)?;
 
-        let (body_async_stream, writer) = async_stream::new_stream(&self.allocator);
+        let (body_async_stream, writer) = async_stream::new_stream(&self.client.allocator);
         message.set_body_stream(Some(body_async_stream));
 
-        let span = request_span!(self, "put_object");
-        span.in_scope(|| debug!(?bucket, ?key, ?params, "new request"));
+        if let Some(len) = content_length {
+            message
+                .add_header(&Header::new("Content-Length", len.to_string()))
+                .map_err(S3RequestError::construction_failure)?;
+        }
+
+        let span = request_span!(self.client, "put_object");
+        span.in_scope(|| debug!(bucket=?self.bucket, ?key, params=?self.params, "new request"));
+
+        let body = self
+            .client
+            .make_simple_http_request(message, MetaRequestType::PutObject, span, |result| {
+                ObjectClientError::ClientError(S3RequestError::ResponseError(result))
+            })?;
+        Ok((body, writer))
+    }
+}
+
+#[async_trait]
+impl PutObjectRequest for S3PutObjectRequest {
+    type ClientError = S3RequestError;
+
+    async fn write(&mut self, slice: &[u8]) -> ObjectClientResult<(), PutObjectError, Self::ClientError> {
+        let (_, writer) = match &mut self.inner {
+            Some(inner) => inner,
+            None => self.inner.insert(self.make_request(None)?),
+        };
+
+        writer
+            .write(slice)
+            .await
+            .map_err(|e| S3RequestError::InternalError(Box::new(e)).into())
+    }
 
-        let body = self.make_simple_http_request(message, MetaRequestType::PutObject, span, |result| {
-            ObjectClientError::ClientError(S3RequestError::ResponseError(result))
-        })?;
+    async fn complete(mut self) -> ObjectClientResult<PutObjectResult, PutObjectError, Self::ClientError> {
+        let body = match self.inner {
+            Some((body, writer)) => {
+                writer
+                    .complete()
+                    .await
+                    .map_err(|e| S3RequestError::InternalError(Box::new(e)))?;
+                body
+            }
+            None => {
+                // If the request has not started yet, this is an empty object and we need to
+                // set Content-Length=0 on the meta request and immediately drop the writer.
+                let (body, writer) = self.make_request(Some(0))?;
+                drop(writer);
+                body
```

Comment:
It's really weird that we don't call `complete` on the writer here. It's because the CRT client never even tries to read the stream if Content-Length is 0? (If that's right, let's clarify the comment)